<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>HCV subtype analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="novel_hcv_subtypes_files/libs/clipboard/clipboard.min.js"></script>
<script src="novel_hcv_subtypes_files/libs/quarto-html/quarto.js"></script>
<script src="novel_hcv_subtypes_files/libs/quarto-html/popper.min.js"></script>
<script src="novel_hcv_subtypes_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="novel_hcv_subtypes_files/libs/quarto-html/anchor.min.js"></script>
<link href="novel_hcv_subtypes_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="novel_hcv_subtypes_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="novel_hcv_subtypes_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="novel_hcv_subtypes_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="novel_hcv_subtypes_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">HCV subtype analysis</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="versions" class="level3">
<h3 class="anchored" data-anchor-id="versions">Versions</h3>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>[1] "R version 4.3.1 (2023-06-16)"</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>        ape   tidyverse      igraph      ggtree      treeio     cowplot 
    "5.7.1"     "2.0.0"     "1.6.0"     "3.9.0"    "1.25.2"     "1.1.1" 
overlapping 
      "2.1" </code></pre>
</div>
</div>
</section>
<section id="hcv-phylogeny" class="level3">
<h3 class="anchored" data-anchor-id="hcv-phylogeny">HCV phylogeny</h3>
<div class="cell">
<div class="cell-output-display">
<p><img src="novel_hcv_subtypes_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="whole-genome-hamming-distances" class="level3">
<h3 class="anchored" data-anchor-id="whole-genome-hamming-distances">Whole-genome Hamming distances</h3>
<div class="cell">
<div class="cell-output-display">
<p><img src="novel_hcv_subtypes_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>For each Pt, largest Wilcox p of all pairwise genetic distances among each Pt and reference sequences of a given genotype:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 2
  seq1  wilcox_p
  &lt;chr&gt;    &lt;dbl&gt;
1 Pt1    0.0286 
2 Pt2    0.0286 
3 Pt3    0.00174</code></pre>
</div>
</div>
</section>
<section id="sliding-window-hamming-distances" class="level3">
<h3 class="anchored" data-anchor-id="sliding-window-hamming-distances">Sliding window Hamming distances</h3>
<div class="cell">
<div class="cell-output-display">
<p><img src="novel_hcv_subtypes_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="novel_hcv_subtypes_files/figure-html/unnamed-chunk-9-2.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Proportion of windows that fell into the between subtype 95% CI for each Pt:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 5
# Groups:   seq1 [3]
  seq1  between_subtype_95         n   tot  prop
  &lt;chr&gt; &lt;chr&gt;                  &lt;int&gt; &lt;int&gt; &lt;dbl&gt;
1 Pt1   Between subtype 95% CI   189   189 1    
2 Pt2   Between subtype 95% CI   168   189 0.889
3 Pt3   Between subtype 95% CI   186   189 0.984</code></pre>
</div>
</div>
</section>
<section id="summary-of-genotype-uncertainty-for-pt-sequences" class="level3">
<h3 class="anchored" data-anchor-id="summary-of-genotype-uncertainty-for-pt-sequences">Summary of genotype uncertainty for Pt sequences</h3>
<div class="cell">
<div class="cell-output-display">
<p><img src="novel_hcv_subtypes_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Proportion of windows that fell into each genotype category (a certain genotype or uncertain):</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 18 × 6
# Groups:   seq1, type [6]
   seq1  type                                geno          n   tot    prop
   &lt;chr&gt; &lt;chr&gt;                               &lt;fct&gt;     &lt;int&gt; &lt;int&gt;   &lt;dbl&gt;
 1 Pt1   Only reference sequences            8            58   189 0.307  
 2 Pt1   Only reference sequences            Other         1   189 0.00529
 3 Pt1   Only reference sequences            Uncertain   130   189 0.688  
 4 Pt1   Pt2 sequence included as genotype 8 8           100   189 0.529  
 5 Pt1   Pt2 sequence included as genotype 8 Other         1   189 0.00529
 6 Pt1   Pt2 sequence included as genotype 8 Uncertain    88   189 0.466  
 7 Pt2   Only reference sequences            8           188   189 0.995  
 8 Pt2   Only reference sequences            Uncertain     1   189 0.00529
 9 Pt2   Pt2 sequence included as genotype 8 8           188   189 0.995  
10 Pt2   Pt2 sequence included as genotype 8 Uncertain     1   189 0.00529
11 Pt3   Only reference sequences            6            10   189 0.0529 
12 Pt3   Only reference sequences            8             6   189 0.0317 
13 Pt3   Only reference sequences            Other         2   189 0.0106 
14 Pt3   Only reference sequences            Uncertain   171   189 0.905  
15 Pt3   Pt2 sequence included as genotype 8 6             8   189 0.0423 
16 Pt3   Pt2 sequence included as genotype 8 8             5   189 0.0265 
17 Pt3   Pt2 sequence included as genotype 8 Other         2   189 0.0106 
18 Pt3   Pt2 sequence included as genotype 8 Uncertain   174   189 0.921  </code></pre>
</div>
</div>
</section>
<section id="summary-of-genotype-uncertainty-for-reference-sequences-and-comparison-to-pt-sequences" class="level3">
<h3 class="anchored" data-anchor-id="summary-of-genotype-uncertainty-for-reference-sequences-and-comparison-to-pt-sequences">Summary of genotype uncertainty for reference sequences (and comparison to Pt sequences)</h3>
<div class="cell">
<div class="cell-output-display">
<p><img src="novel_hcv_subtypes_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>IQR of proportion of windows with uncertain genotype for reference sequences:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 3
  type                   q25   q75
  &lt;fct&gt;                &lt;dbl&gt; &lt;dbl&gt;
1 "Existing\ngenotype" 0     0.511
2 "New\ngenotype"      0.746 0.921</code></pre>
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>